### DISCUSSION TITLE
High Performance Product Storage

### DISCUSSION THREAD

**Discussion by MIkram7** [woocommerce/woocommerce#43447] [General]
Are there any plans to optimise the way products/tags/attributes are stored in WooCommerce? Similar to the amazing achievements of HPOS.

We're desperately in need of a revamp for products. We have a very active store with 3500+ products with 12,000+ product attributes. Happy to help contribute to develop this feature.

**Comment by MIkram7** [comment]
Some more context:

We have 3500+ products. On the front end, there are three attributes a customer can filter on. 

**Let's consider an example. Let's say:**

- Attribute 1 (Colour) has 22 options.
- Attribute 2 (Length) has 51 options. 
- Attribute 3 (Width) has 44 options.

This could result in query parameter filters such as:

- /product-category/my-category/?filter_length=54,53&query_type_length=or&filter_colour=black&query_type_colour=or&filter_width=27&query_type_width=or
- /product-category/my-second-category/?filter_colour=black,blue&query_type_colour=or&filter_width=27,28,29&query_type_width=or
- /product-category/my-category/?filter_colour=black&query_type_colour=or&filter_length=54,53&query_type_length=or&filter_width=27&query_type_width=or
- /shop/?filter_colour=black&query_type_colour=or

If we remove or swap a single parameter from the query, it is a different result and therefore has a different cache. 

If we were to just try and brute-force cache each query param, with the given combinations above, we'd have ~49k+ combinations. It's not feasible. Add into that the complexity of the ordering of the parameters in the URL and it's quickly proving to be a disaster. 

Stripping query params is also not an option as it will serve the same cached page for all parameters.

We believe caching, although helpful, is only a bandaid and does not address the underlying issue. Some of our tables (e.g. _posts) have millions of rows and without sufficient indexing, it's a huge performance blow.

**Reply by lkraav** [reply]
Yes. We had to build a custom search index on top of WP posts/meta model to make performance manageable. This was years ago, but is still in production. Today, I would look into whether ElasticPress can help, first.
